### ISSUE TITLE
Firebase-queue for Cloud Functions

### ISSUE THREAD

**Issue by jaufgang** [firebase/firebase-queue#115]
I am basing this issue on a [question](http://stackoverflow.com/questions/42724957/does-cloud-functions-for-firebase-make-firebase-queue-obsolete) I came across on StackOverflow regarding whether or not cloud functions render firebase-queue obsolete, and [my answer](http://stackoverflow.com/questions/42724957/does-cloud-functions-for-firebase-make-firebase-queue-obsolete/42755440#42755440) to that question.  Also, I noticed that the Readme.md has been changed to point developers away from this library in favour of cloud functions.

Firebase-queue provides much more than a way for tasks to listen to firebase and launch tasks. It provides a protocol for communication between parties assigning and and responding to task requests, coordinating retries of failed tasks, reporting progress, state and additional metadata, and task chaining.

I think it would be useful for Firebase to develop a firebase-functions-queue package that extends the firebase-functions package and allows you to write a firebase cloud function with the same signature as firebase-queue. Something like this:

```
const functions = require('firebase-functions');
const functions-queue = require('firebase-functions-queue'); //extends firebase-functions with onQueue function

admin.initializeApp(functions.config().firebase);

functions.database.ref('/queue').onQueue(options,function(data,progress,resolve,reject){
    ...
    })
```
This new package would work just like firebase-queue and use the same metadata and options, except it wouldn't need to deal with managing multiple worker processes since cloud functions already do this automatically and seamlessly.

This would have the following advantages:

- Would allow firebase developers to use a standard way of assigning jobs to queues, monitor progress, failure, etc.
- An app could assign a job to a queue and not care if it was being handled by a cloud function or a different environment.
- A developer could take an existing queue and move it from their node server to cloud functions without changing the client app.
- It could even allow task chaining where one task could be handled by a cloud function and another by a different server within the same job.

**Comment by katowulf** [comment]
@drtriumph @nicolasgarnier fyi

**Comment by cha211** [comment]
another advantage is having a queue running on a local server that uses platform specific libraries that aren't supported on Cloud Functions

**Comment by prescottprue** [comment]
I agree with this. Have ended up just doing my own implementation instead of using queue

**Comment by alexanvl** [comment]
This is interesting to me.  I've done a few projects with firebase now and really like firebase-queue.  I recently learned about functions and am curious how I can use it similarly and in combination with pub/sub to eliminate a dependency on amqp.

**Comment by IncognitoJam** [comment]
I am also interested in something like this and am sad to have not heard anything from the developers of this repository. `firebase-queue` offers a lot more than firebase functions and being able to deploy them to Firebase would be great.

**Comment by EECOLOR** [comment]
@jaufgang I just released a trimmed down version of firebase queue: https://github.com/kaliberjs/firebase-queue

It is a lot more lightweight, has less moving parts, no dependencies and a more modern approach (promises).

I am not sure what would be needed to support Firebase Functions, but if you look at the code you will see that it is probably pretty doable. I personally have no experience with Firebase Functions, but I think if there is enough demand and interest we can pull something off.
